lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU

```diff
@@ -99,19 +160,91 @@ fbGetPMVoltage(
     fTermMin := 0,
     fReal => fVoltage);
 
-// Power meter calibration
-fCalibBase := (fVoltage + fCalibRelOffset) * fCalibRelRatio;
-fCalibMJ := fCalibBase * fCalibMJRatio;
+IF R_TRIG(CLK:=(eEnumGet <> E_PPM_States.OUT OR stYAxisState <> 0)) THEN
+    // Dump most recent readings
+    IF udBackgroundVoltageBufferIndex > 5 THEN
+        FOR i := (udBackgroundVoltageBufferIndex - 5) TO udBackgroundVoltageBufferIndex DO
+            fBackgroundVoltageSum := fBackgroundVoltageSum - fBackgroundVoltageBuffer[i MOD 100];
+        END_FOR
+        udBackgroundVoltageBufferIndex := udBackgroundVoltageBufferIndex - 5;
+        IF udBackgroundVoltageBufferIndex <> 1 AND BACKGROUND_MODE = Passive THEN
+            fBackgroundVoltage := fBackgroundVoltageSum / MAX((udBackgroundVoltageBufferIndex - 1), 100);
+        END_IF
+    END_IF
+    fBackgroundVoltageAcc := 0;
+    fBackgroundVoltageSum := 0;
+    tonBackgroundAutoCollecting(IN := FALSE);
+    udBackgroundVoltageBufferIndex := 1;
+ELSIF eEnumGet = E_PPM_States.OUT AND stYAxisState = 0 THEN
```

Comment:
Following along: if we didn't just start a move and we're standing still in the OUT state, wait 1 second and then sum the accumulated background voltages?

Reply:
Yes. I'm using fBackgroundVoltageAcc so to accumulate every cycle so that the array is a reasonable size and to make dumping values when a move is started simpler.